src/autoschedulers/adams2019/FunctionDAG.cpp

```diff
@@ -549,12 +549,48 @@ void FunctionDAG::Edge::expand_footprint(const Span *consumer_loop, Span *produc
     }
 }
 
-FunctionDAG::FunctionDAG(const vector<Function> &outputs, const MachineParams &params, const Target &target) {
+FunctionDAG::FunctionDAG(const vector<Function> &outputs_arg, const MachineParams &params, const Target &target) {
+    // Bounds inference can include image loads that are to a single
+    // location in returned bounds expressions. The idea is these are
+    // easily computed at runtime entry to the compiled code. However
+    // the values are not available to the autoscheduler and hence
+    // these cannot be used. Thus we rewrite such Image calls to
+    // Extern calls here to hide them from bounds inference.
+    class RemoveImageLoads : public IRMutator {
+        using IRMutator::visit;
+
+        Expr visit(const Call *op) override {
+            if (op->call_type == Call::Image) {
+                vector<Expr> new_args(op->args.size());
+                for (size_t i = 0; i < op->args.size(); i++) {
+                    const Expr &old_arg = op->args[i];
+                    Expr new_arg = mutate(old_arg);
+                    new_args[i] = std::move(new_arg);
+                }
+
+                return Call::make(op->type, op->name, new_args, Call::Extern, op->func, op->value_index, op->image, op->param);
+            } else {
+                return IRMutator::visit(op);
+            }
+        }
+    } remove_image_loads;
+
     map<string, Function> env;
-    for (const Function &o : outputs) {
+    for (const Function &o : outputs_arg) {
         populate_environment(o, env);
     }
 
+    std::vector<Function> outputs;
```

Comment:
I think this is going to mess up the featurization, which happens at the end of this constructor, because image loads are treated differently to extern calls. Can this mutation not just been done only for the things passed to compute_function_value_bounds and boxes_required? Another option would be making two dags (i.e. keeping the original after the deep copy)

Reply:
Certainly easy enough to change the code. Any suggestions for how to test this sort of thing? Either using existing tests or adding one.